File: contracts/remote/v1/conformance-go/conformance/endpoint_envelope_test.go
package conformance_test

import (
	"bytes"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"strconv"
	"testing"

	"github.com/waifucave/discord-waifus/contracts/remote/v1/conformance-go/internal/pairing"
	"github.com/waifucave/discord-waifus/contracts/remote/v1/conformance-go/internal/vectors"
)

type endpointContextVector struct {
	NegotiatedMinor                 int    `json:"negotiatedMinor"`
	PairIDB64                       string `json:"pairIdB64"`
	SenderRole                      int    `json:"senderRole"`
	ReceiverRole                    int    `json:"receiverRole"`
	HostInstallationBundleHashB64   string `json:"hostInstallationBundleHashB64"`
	RemoteInstallationBundleHashB64 string `json:"remoteInstallationBundleHashB64"`
	HostTrustEpoch                  string `json:"hostTrustEpoch"`
	RemoteTrustEpoch                string `json:"remoteTrustEpoch"`
	EndpointEpoch                   string `json:"endpointEpoch"`
}

type endpointCandidateVector struct {
	Kind       int    `json:"kind"`
	Family     int    `json:"family"`
	AddressB64 string `json:"addressB64"`
	Port       int    `json:"port"`
	Priority   int64  `json:"priority"`
}

type endpointValueVector struct {
	Version              int                       `json:"version"`
	EndpointEpoch        string                    `json:"endpointEpoch"`
	ConnectionGeneration string                    `json:"connectionGeneration"`
	Candidates           []endpointCandidateVector `json:"candidates"`
}

type endpointEnvelopeVector struct {
	Name                string                `json:"name"`
	Context             endpointContextVector `json:"context"`
	Value               endpointValueVector   `json:"value"`
	PlaintextB64        string                `json:"plaintextB64"`
	NonceB64            string                `json:"nonceB64"`
	AssociatedDataB64   string                `json:"associatedDataB64"`
	CiphertextB64       string                `json:"ciphertextB64"`
	CiphertextSHA256B64 string                `json:"ciphertextSha256B64"`
}

type endpointRejectionVector struct {
	Name          string                `json:"name"`
	PlaintextB64  string                `json:"plaintextB64"`
	Context       endpointContextVector `json:"context"`
	CiphertextB64 string                `json:"ciphertextB64"`
	Approved      bool                  `json:"approved"`
	ErrorCode     string                `json:"errorCode"`
}

type endpointFixture struct {
	Keys struct {
		HostToRemoteKeyB64 string `json:"hostToRemoteKeyB64"`
		RemoteToHostKeyB64 string `json:"remoteToHostKeyB64"`
	} `json:"keys"`
	Envelopes    []endpointEnvelopeVector `json:"envelopes"`
	EpochAdvance struct {
		Context       endpointContextVector `json:"context"`
		Value         endpointValueVector   `json:"value"`
		CiphertextB64 string                `json:"ciphertextB64"`
	} `json:"epochAdvance"`
	Boundary struct {
		KeyB64                  string `json:"keyB64"`
		NonceB64                string `json:"nonceB64"`
		AssociatedDataB64       string `json:"associatedDataB64"`
		PlaintextB64            string `json:"plaintextB64"`
		CiphertextB64           string `json:"ciphertextB64"`
		OverLimitPlaintextB64   string `json:"overLimitPlaintextB64"`
		MaximumValidRecordBytes int    `json:"maximumValidRecordBytes"`
	} `json:"boundary"`
	Rejections struct {
		Plaintext []endpointRejectionVector `json:"plaintext"`
		Envelopes []endpointRejectionVector `json:"envelopes"`
	} `json:"rejections"`
}

func endpointFixtureBytes(t *testing.T) []byte {
	t.Helper()
	value, err := os.ReadFile(filepath.Join("..", "..", "fixtures", "crypto", "endpoint-envelope-v1.json"))
	if err != nil {
		t.Fatalf("read endpoint fixture: %v", err)
	}
	return value
}

func decodeEndpointFixture(t *testing.T) endpointFixture {
	t.Helper()
	var value endpointFixture
	if err := json.Unmarshal(endpointFixtureBytes(t), &value); err != nil {
		t.Fatalf("decode endpoint fixture: %v", err)
	}
	return value
}

func endpointUint64(t *testing.T, value string) uint64 {
	t.Helper()
	parsed, err := strconv.ParseUint(value, 10, 64)
	if err != nil || strconv.FormatUint(parsed, 10) != value {
		t.Fatalf("parse endpoint uint64 %q: %v", value, err)
	}
	return parsed
}

func endpointContext(t *testing.T, value endpointContextVector) pairing.EndpointContext {
	t.Helper()
	if value.NegotiatedMinor < 0 || value.NegotiatedMinor > 65535 || value.SenderRole < 0 || value.SenderRole > 255 || value.ReceiverRole < 0 || value.ReceiverRole > 255 {
		t.Fatal("endpoint fixture context integer is out of range")
	}
	return pairing.EndpointContext{
		NegotiatedMinor:              uint16(value.NegotiatedMinor),
		PairID:                       decodeBase64URL(t, value.PairIDB64),
		SenderRole:                   uint8(value.SenderRole),
		ReceiverRole:                 uint8(value.ReceiverRole),
		HostInstallationBundleHash:   decodeBase64URL(t, value.HostInstallationBundleHashB64),
		RemoteInstallationBundleHash: decodeBase64URL(t, value.RemoteInstallationBundleHashB64),
		HostTrustEpoch:               endpointUint64(t, value.HostTrustEpoch),
		RemoteTrustEpoch:             endpointUint64(t, value.RemoteTrustEpoch),
		EndpointEpoch:                endpointUint64(t, value.EndpointEpoch),
	}
}

func endpointValue(t *testing.T, value endpointValueVector) pairing.EndpointGeneration {
	t.Helper()
	if value.Version < 0 || value.Version > 255 {
		t.Fatal("endpoint fixture version is out of range")
	}
	candidates := make([]pairing.EndpointCandidate, len(value.Candidates))
	for index, candidate := range value.Candidates {
		if candidate.Kind < 0 || candidate.Kind > 255 || candidate.Family < 0 || candidate.Family > 255 ||
			candidate.Port < 0 || candidate.Port > 65535 || candidate.Priority < 0 || candidate.Priority > 4294967295 {
			t.Fatalf("endpoint candidate %d integer is out of range", index)
		}
		candidates[index] = pairing.EndpointCandidate{
			Kind: uint8(candidate.Kind), Family: uint8(candidate.Family),
			Address: decodeBase64URL(t, candidate.AddressB64), Port: uint16(candidate.Port), Priority: uint32(candidate.Priority),
		}
	}
	return pairing.EndpointGeneration{
		Version: uint8(value.Version), EndpointEpoch: endpointUint64(t, value.EndpointEpoch),
		ConnectionGeneration: endpointUint64(t, value.ConnectionGeneration), Candidates: candidates,
	}
}

func endpointKeys(t *testing.T, fixture endpointFixture) pairing.EndpointDirectionKeys {
	t.Helper()
	return pairing.EndpointDirectionKeys{
		HostToRemote: decodeBase64URL(t, fixture.Keys.HostToRemoteKeyB64),
		RemoteToHost: decodeBase64URL(t, fixture.Keys.RemoteToHostKeyB64),
	}
}

func endpointErrorCode(t *testing.T, err error) string {
	t.Helper()
	var protocolError *pairing.EndpointError
	if !errors.As(err, &protocolError) {
		t.Fatalf("expected EndpointError, got %T: %v", err, err)
	}
	return protocolError.Code
}

func TestEndpointFixtureMatchesIndependentGenerator(t *testing.T) {
	expected, err := vectors.BuildEndpointEnvelopeV1JSON()
	if err != nil {
		t.Fatalf("build independent endpoint fixture: %v", err)
	}
	if !bytes.Equal(endpointFixtureBytes(t), expected) {
		t.Fatal("committed endpoint fixture differs from independent Go generator")
	}
}

func TestEndpointEnvelopesBothDirections(t *testing.T) {
	fixture := decodeEndpointFixture(t)
	keys := endpointKeys(t, fixture)
	if len(fixture.Envelopes) != 3 {
		t.Fatalf("expected three endpoint vectors, got %d", len(fixture.Envelopes))
	}
	for _, vector := range fixture.Envelopes {
		context := endpointContext(t, vector.Context)
		value := endpointValue(t, vector.Value)
		encrypted, err := pairing.EncryptEndpointEnvelope(keys, context, value, true)
		if err != nil {
			t.Fatalf("encrypt %s: %v", vector.Name, err)
		}
		comparisons := []struct {
			name string
			got  []byte
			want string
		}{
			{"plaintext", encrypted.Plaintext, vector.PlaintextB64},
			{"nonce", encrypted.Nonce, vector.NonceB64},
			{"associated data", encrypted.AssociatedData, vector.AssociatedDataB64},
			{"ciphertext", encrypted.Ciphertext, vector.CiphertextB64},
			{"ciphertext hash", pairing.Hash(encrypted.Ciphertext), vector.CiphertextSHA256B64},
		}
		for _, comparison := range comparisons {
			if pairing.B64(comparison.got) != comparison.want {
				t.Errorf("%s %s differs", vector.Name, comparison.name)
			}
		}
		decoded, err := pairing.DecodeEndpointPlaintext(encrypted.Plaintext)
		if err != nil || !reflect.DeepEqual(decoded, value) {
			t.Errorf("decode %s plaintext: %v", vector.Name, err)
		}
		minor := context.NegotiatedMinor
		decodedContext, err := pairing.DecodeEndpointAssociatedData(encrypted.AssociatedData, &minor)
		if err != nil || !reflect.DeepEqual(decodedContext, context) {
			t.Errorf("decode %s associated data: %v", vector.Name, err)
		}
		decrypted, err := pairing.DecryptEndpointEnvelope(keys, context, encrypted.Ciphertext, true)
		if err != nil || !reflect.DeepEqual(decrypted, value) {
			t.Errorf("decrypt %s envelope: %v", vector.Name, err)
		}
	}
}

func TestEndpointPlaintextAndEnvelopeRejections(t *testing.T) {
	fixture := decodeEndpointFixture(t)
	for _, rejection := range fixture.Rejections.Plaintext {
		_, err := pairing.DecodeEndpointPlaintext(decodeBase64URL(t, rejection.PlaintextB64))
		if err == nil || endpointErrorCode(t, err) != rejection.ErrorCode {
			t.Errorf("plaintext rejection %s returned %v", rejection.Name, err)
		}
	}
	keys := endpointKeys(t, fixture)
	for _, rejection := range fixture.Rejections.Envelopes {
		_, err := pairing.DecryptEndpointEnvelope(
			keys,
			endpointContext(t, rejection.Context),
			decodeBase64URL(t, rejection.CiphertextB64),
			rejection.Approved,
		)
		if err == nil || endpointErrorCode(t, err) != rejection.ErrorCode {
			t.Errorf("envelope rejection %s returned %v", rejection.Name, err)
		}
	}
}

func TestEndpointRawAEADBoundary(t *testing.T) {
	fixture := decodeEndpointFixture(t)
	key := decodeBase64URL(t, fixture.Boundary.KeyB64)
	nonce := decodeBase64URL(t, fixture.Boundary.NonceB64)
	associatedData := decodeBase64URL(t, fixture.Boundary.AssociatedDataB64)
	plaintext := decodeBase64URL(t, fixture.Boundary.PlaintextB64)
	ciphertext, err := pairing.EncryptEndpointAEAD(key, nonce, associatedData, plaintext)
	if err != nil {
		t.Fatalf("encrypt endpoint boundary: %v", err)
	}
	if len(plaintext) != pairing.EndpointPlaintextMax || len(ciphertext) != pairing.EndpointCiphertextMax || pairing.B64(ciphertext) != fixture.Boundary.CiphertextB64 {
		t.Fatal("endpoint AEAD boundary differs")
	}
	decrypted, err := pairing.DecryptEndpointAEAD(key, nonce, associatedData, ciphertext)
	if err != nil || !bytes.Equal(decrypted, plaintext) {
		t.Fatalf("decrypt endpoint boundary: %v", err)
	}
	_, err = pairing.EncryptEndpointAEAD(key, nonce, associatedData, decodeBase64URL(t, fixture.Boundary.OverLimitPlaintextB64))
	if err == nil || endpointErrorCode(t, err) != "plaintext_too_large" {
		t.Fatalf("over-limit endpoint plaintext returned %v", err)
	}
	if _, err := pairing.DecodeEndpointPlaintext(plaintext); err == nil || endpointErrorCode(t, err) != "invalid_canonical_cbor" {
		t.Fatalf("raw boundary was accepted as endpoint CBOR: %v", err)
	}
}

func TestEndpointReceiveStateRecoveryConflictAndRollback(t *testing.T) {
	fixture := decodeEndpointFixture(t)
	keys := endpointKeys(t, fixture)
	first := fixture.Envelopes[0]
	context := endpointContext(t, first.Context)
	ciphertext := decodeBase64URL(t, first.CiphertextB64)
	state, err := pairing.NewEndpointReceiveState()
	if err != nil {
		t.Fatalf("create endpoint receive state: %v", err)
	}
	prepared, err := state.Prepare(keys, context, ciphertext, true)
	if err != nil || prepared.Status != "prepared" || state.Snapshot().Phase != "prepared" {
		t.Fatalf("prepare endpoint: %+v %v", prepared, err)
	}
	restarted, err := pairing.NewEndpointReceiveState(state.Snapshot())
	if err != nil {
		t.Fatalf("restore endpoint state: %v", err)
	}
	resumed, err := restarted.Prepare(keys, context, ciphertext, true)
	if err != nil || resumed.Status != "resume_prepared" {
		t.Fatalf("resume endpoint: %+v %v", resumed, err)
	}
	if err := state.MarkApplied(context.EndpointEpoch, prepared.CiphertextSHA256); err != nil {
		t.Fatalf("mark endpoint applied: %v", err)
	}
	idempotent, err := state.Prepare(keys, context, ciphertext, true)
	if err != nil || idempotent.Status != "already_applied" {
		t.Fatalf("idempotent applied endpoint: %+v %v", idempotent, err)
	}
	if _, err := state.Prepare(keys, context, ciphertext, false); err == nil || endpointErrorCode(t, err) != "unapproved_sender" {
		t.Fatalf("unapproved idempotent endpoint returned %v", err)
	}
	conflict := append([]byte(nil), ciphertext...)
	conflict[0] ^= 1
	if _, err := state.Prepare(keys, context, conflict, true); err == nil || endpointErrorCode(t, err) != "epoch_conflict" {
		t.Fatalf("same-epoch conflict returned %v", err)
	}
	advancedContext := endpointContext(t, fixture.EpochAdvance.Context)
	advanced, err := state.Prepare(keys, advancedContext, decodeBase64URL(t, fixture.EpochAdvance.CiphertextB64), true)
	if err != nil || advanced.Status != "prepared" {
		t.Fatalf("advance endpoint epoch: %+v %v", advanced, err)
	}
	if _, err := state.Prepare(keys, context, ciphertext, true); err == nil || endpointErrorCode(t, err) != "epoch_rollback" {
		t.Fatalf("endpoint rollback returned %v", err)
	}
}

File: tests/endpointEnvelope.test.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import { describe, expect, it } from "vitest";
import {
  ENDPOINT_CIPHERTEXT_MAX_BYTES,
  ENDPOINT_PLAINTEXT_MAX_BYTES,
  EndpointReceiveStateV1,
  decodeEndpointAssociatedDataV1,
  decodeEndpointPlaintextV1,
  decryptEndpointAeadPayloadV1,
  decryptEndpointEnvelopeV1,
  encodeEndpointAssociatedDataV1,
  encodeEndpointPlaintextV1,
  encryptEndpointAeadPayloadV1,
  encryptEndpointEnvelopeV1,
  endpointNonceV1,
  type EndpointDirectionKeysV1,
  type EndpointEnvelopeContextV1,
  type EndpointGenerationV1
} from "../src/shared/endpointEnvelope.js";
import {
  createEndpointEnvelopeV1Fixture,
  serializeEndpointEnvelopeFixture
} from "../src/shared/endpointEnvelopeContract.js";

type JsonRecord = Record<string, unknown>;

function record(value: unknown): JsonRecord {
  expect(value).not.toBeNull();
  expect(typeof value).toBe("object");
  expect(Array.isArray(value)).toBe(false);
  return value as JsonRecord;
}

function text(value: unknown): string {
  expect(typeof value).toBe("string");
  return value as string;
}

function integer(value: unknown): number {
  expect(typeof value).toBe("number");
  expect(Number.isInteger(value)).toBe(true);
  return value as number;
}

function bytes(value: unknown): Buffer {
  return Buffer.from(text(value), "base64url");
}

function endpointContext(value: JsonRecord): EndpointEnvelopeContextV1 {
  return {
    negotiatedMinor: integer(value.negotiatedMinor),
    pairId: bytes(value.pairIdB64),
    senderRole: integer(value.senderRole) as 1 | 2,
    receiverRole: integer(value.receiverRole) as 1 | 2,
    hostInstallationBundleHash: bytes(value.hostInstallationBundleHashB64),
    remoteInstallationBundleHash: bytes(value.remoteInstallationBundleHashB64),
    hostTrustEpoch: BigInt(text(value.hostTrustEpoch)),
    remoteTrustEpoch: BigInt(text(value.remoteTrustEpoch)),
    endpointEpoch: BigInt(text(value.endpointEpoch))
  };
}

function endpointRecord(value: JsonRecord): EndpointGenerationV1 {
  return {
    version: 1,
    endpointEpoch: BigInt(text(value.endpointEpoch)),
    connectionGeneration: BigInt(text(value.connectionGeneration)),
    candidates: (value.candidates as JsonRecord[]).map((candidate) => ({
      kind: integer(candidate.kind) as 1 | 2 | 3,
      family: integer(candidate.family) as 4 | 6,
      address: bytes(candidate.addressB64),
      port: integer(candidate.port),
      priority: integer(candidate.priority)
    }))
  };
}

function directionKeys(value: JsonRecord): EndpointDirectionKeysV1 {
  return {
    hostToRemoteKey: bytes(value.hostToRemoteKeyB64),
    remoteToHostKey: bytes(value.remoteToHostKeyB64)
  };
}

describe("endpoint envelope V1 public crypto vectors", () => {
  it("recreates the committed fixture byte-for-byte", async () => {
    const fixturePath = path.join(
      process.cwd(),
      "contracts",
      "remote",
      "v1",
      "fixtures",
      "crypto",
      "endpoint-envelope-v1.json"
    );
    const actual = await readFile(fixturePath, "utf8");
    expect(actual).toBe(serializeEndpointEnvelopeFixture(createEndpointEnvelopeV1Fixture()));
  });

  it("pins canonical endpoint CBOR, nonce, associated data, and both direction keys", () => {
    const fixture = record(createEndpointEnvelopeV1Fixture());
    const keys = directionKeys(record(fixture.keys));
    for (const vectorValue of fixture.envelopes as JsonRecord[]) {
      const vector = record(vectorValue);
      const context = endpointContext(record(vector.context));
      const value = endpointRecord(record(vector.value));
      const encrypted = encryptEndpointEnvelopeV1({
        directionKeys: keys,
        context,
        value,
        approved: true
      });

      expect(encodeEndpointPlaintextV1(value).toString("base64url"))
        .toBe(text(vector.plaintextB64));
      expect(endpointNonceV1(context.endpointEpoch).toString("base64url"))
        .toBe(text(vector.nonceB64));
      expect(encodeEndpointAssociatedDataV1(context).toString("base64url"))
        .toBe(text(vector.associatedDataB64));
      expect(encrypted.ciphertext.toString("base64url")).toBe(text(vector.ciphertextB64));
      expect(decodeEndpointPlaintextV1(encrypted.plaintext)).toEqual(value);
      expect(decodeEndpointAssociatedDataV1(encrypted.associatedData, context.negotiatedMinor))
        .toEqual(context);
      expect(decryptEndpointEnvelopeV1({
        directionKeys: keys,
        context,
        ciphertext: encrypted.ciphertext,
        approved: true
      })).toEqual(value);
    }
  });

  it("rejects malformed, unsafe, duplicate, or unsorted candidate records", () => {
    const fixture = record(createEndpointEnvelopeV1Fixture());
    for (const rejectionValue of record(fixture.rejections).plaintext as JsonRecord[]) {
      const rejection = record(rejectionValue);
      expect(
        () => decodeEndpointPlaintextV1(bytes(rejection.plaintextB64)),
        text(rejection.name)
      ).toThrow(new RegExp(text(rejection.errorCode)));
    }
  });

  it("rejects tamper, wrong direction, context substitution, and unapproved exchange", () => {
    const fixture = record(createEndpointEnvelopeV1Fixture());
    const keys = directionKeys(record(fixture.keys));
    for (const rejectionValue of record(fixture.rejections).envelopes as JsonRecord[]) {
      const rejection = record(rejectionValue);
      expect(() => decryptEndpointEnvelopeV1({
        directionKeys: keys,
        context: endpointContext(record(rejection.context)),
        ciphertext: bytes(rejection.ciphertextB64),
        approved: rejection.approved as boolean
      }), text(rejection.name)).toThrow(new RegExp(text(rejection.errorCode)));
    }
  });

  it("pins the raw AEAD 1,184/1,200-byte ceiling without treating padding as endpoint CBOR", () => {
    const fixture = record(createEndpointEnvelopeV1Fixture());
    const boundary = record(fixture.boundary);
    const key = bytes(boundary.keyB64);
    const nonce = bytes(boundary.nonceB64);
    const associatedData = bytes(boundary.associatedDataB64);
    const plaintext = bytes(boundary.plaintextB64);
    const ciphertext = encryptEndpointAeadPayloadV1(key, nonce, associatedData, plaintext);

    expect(plaintext).toHaveLength(ENDPOINT_PLAINTEXT_MAX_BYTES);
    expect(ciphertext).toHaveLength(ENDPOINT_CIPHERTEXT_MAX_BYTES);
    expect(ciphertext.toString("base64url")).toBe(text(boundary.ciphertextB64));
    expect(decryptEndpointAeadPayloadV1(key, nonce, associatedData, ciphertext)).toEqual(plaintext);
    expect(() => encryptEndpointAeadPayloadV1(
      key,
      nonce,
      associatedData,
      Buffer.alloc(ENDPOINT_PLAINTEXT_MAX_BYTES + 1)
    )).toThrow(/plaintext_too_large/);
    expect(() => decodeEndpointPlaintextV1(plaintext)).toThrow(/invalid_canonical_cbor/);
  });

  it("persists prepared/applied receive state and rejects conflict or rollback", () => {
    const fixture = record(createEndpointEnvelopeV1Fixture());
    const keys = directionKeys(record(fixture.keys));
    const first = record((fixture.envelopes as JsonRecord[])[0]);
    const context = endpointContext(record(first.context));
    const ciphertext = bytes(first.ciphertextB64);
    const state = new EndpointReceiveStateV1();

    const prepared = state.prepare({
      directionKeys: keys,
      context,
      ciphertext,
      approved: true
    });
    expect(prepared.status).toBe("prepared");
    expect(state.snapshot().phase).toBe("prepared");

    const resumed = new EndpointReceiveStateV1(state.snapshot()).prepare({
      directionKeys: keys,
      context,
      ciphertext,
      approved: true
    });
    expect(resumed.status).toBe("resume_prepared");
    state.markApplied(context.endpointEpoch, prepared.ciphertextSha256);
    expect(state.snapshot().phase).toBe("applied");
    expect(state.prepare({ directionKeys: keys, context, ciphertext, approved: true }).status)
      .toBe("already_applied");
    expect(() => state.prepare({
      directionKeys: keys,
      context,
      ciphertext,
      approved: false
    })).toThrow(/unapproved_sender/);

    const conflict = Buffer.from(ciphertext);
    conflict[0] ^= 1;
    expect(() => state.prepare({
      directionKeys: keys,
      context,
      ciphertext: conflict,
      approved: true
    })).toThrow(/epoch_conflict/);

    const advanced = record(fixture.epochAdvance);
    const advancedResult = state.prepare({
      directionKeys: keys,
      context: endpointContext(record(advanced.context)),
      ciphertext: bytes(advanced.ciphertextB64),
      approved: true
    });
    expect(advancedResult.status).toBe("prepared");
    expect(() => state.prepare({
      directionKeys: keys,
      context,
      ciphertext,
      approved: true
    })).toThrow(/epoch_rollback/);
  });
});
